decoder.go

```diff
@@ -198,54 +198,23 @@ func DecodeFloat64(bz []byte) (f float64, n int, err error) {
 // undefined.
 // TODO return error if behavior is undefined.
 func DecodeTime(bz []byte) (t time.Time, n int, err error) {
-
+	t = zeroTime
 	// TODO: This is a temporary measure until we support MarshalAmino/UnmarshalAmino.
 	// Basically, MarshalAmino on time should return a struct.
 	// This is how that struct would be encoded.
-
-	{ // Decode field number 1 and Typ3 (8Byte).
-		var fieldNum, typ, _n = uint32(0), Typ3(0x00), int(0)
-		fieldNum, typ, _n, err = decodeFieldNumberAndTyp3(bz)
-		if slide(&bz, &n, _n) && err != nil {
-			return
-		}
-		if fieldNum != 1 {
-			err = fmt.Errorf("expected field number 1, got %v", fieldNum)
-			return
-		}
-		if typ != Typ3_8Byte {
-			err = fmt.Errorf("expected Typ3 bytes <8Bytes> for time field #1, got %v", typ)
-			return
-		}
-	}
-	// Actually read the Int64.
-	var sec, _n = int64(0), int(0)
-	sec, _n, err = DecodeInt64(bz)
-	if slide(&bz, &n, _n) && err != nil {
+	var sec int64
+	if len(bz) == 0 {
 		return
 	}
-
-	{ // Decode field number 2 and Typ3 (4Byte).
-		var fieldNum, typ, _n = uint32(0), Typ3(0x00), int(0)
-		fieldNum, typ, _n, err = decodeFieldNumberAndTyp3(bz)
-		if slide(&bz, &n, _n) && err != nil {
-			return
-		}
-		if fieldNum != 2 {
-			err = fmt.Errorf("expected field number 2, got %v", fieldNum)
-			return
-		}
-		if typ != Typ3_4Byte {
-			err = fmt.Errorf("expected Typ3 bytes <4Byte> for time field #2, got %v", typ)
-			return
-		}
+	sec, n, err = decodeSeconds(&bz)
+	if err != nil {
+		return
 	}
-	// Actually read the Int32.
-	var nsec = int32(0)
-	nsec, _n, err = DecodeInt32(bz)
-	if slide(&bz, &n, _n) && err != nil {
+	nsec, err := decodeNanos(&bz, &n)
```

Comment:
The implicit return values make it hard to reason about what is returned at what point. Here for example is the returned `err` set correctly? If possible we should avoid them, good reasons to use implicit returns is for access in `defer`s.

Reply:
Yes, returns the correct err. But I agree with with the implicit returns. I did not want to change the public method here for the sake of consistency: all methods in this file have named return values/implicit returns.

Reply:
Understood, changing those won't affect anything on the caller side tho.